File: bench/s01_bounds_check.c
/* s01_bounds_check.c — bounds check on/off overhead
 * Compares "with bounds check" vs "without bounds check" array access.
 * N=10000000 elements, R=10 rounds, pseudo-random index pattern.
 * Init: arr[i]=i, idx[i]=(i*1103515245+12345)%N (well-defined via uint32).
 * With check:    if (ii >= 0 && ii < N) { sum += arr[ii]; }
 * Without check: sum += arr[idx[i]];
 * Anti-fold: __asm__ volatile in bench loop bodies; bench returns sum. */
#include <stdint.h>
#include <stdio.h>
#include <time.h>

enum { N = 10000000, R = 10 };

static int32_t arr[N];
static int32_t idx[N];

static int32_t bench_with_check(void) {
  int32_t sum = 0;
  int32_t r = 0;
  while (r < R) {
    int32_t i = 0;
    while (i < N) {
      int32_t ii = idx[i];
      if (ii >= 0 && ii < N) {
        sum += arr[ii];
      }
      i = i + 1;
    }
    r = r + 1;
    /* B-CMP: make sum opaque per round to prevent DCE / folding. */
    __asm__ volatile("" : "+r"(sum) : : "memory");
  }
  return sum;
}

static int32_t bench_without_check(void) {
  int32_t sum = 0;
  int32_t r = 0;
  while (r < R) {
    int32_t i = 0;
    while (i < N) {
      sum += arr[idx[i]];
      i = i + 1;
    }
    r = r + 1;
    __asm__ volatile("" : "+r"(sum) : : "memory");
  }
  return sum;
}

int main(void) {
  int32_t i = 0;
  while (i < N) {
    arr[i] = i;
    idx[i] = (int32_t)(((uint32_t)i * 1103515245u + 12345u) % (uint32_t)N);
    i = i + 1;
  }
  /* B-CMP: memory barrier prevents the compiler from assuming arr/idx
   * contents are known at compile time during the bench loops. */
  __asm__ volatile("" : : : "memory");

  clock_t t1 = clock();
  int32_t s1 = bench_with_check();
  clock_t t2 = clock();
  int32_t s2 = bench_without_check();
  clock_t t3 = clock();

  double ms1 = (double)(t2 - t1) / CLOCKS_PER_SEC * 1000.0;
  double ms2 = (double)(t3 - t2) / CLOCKS_PER_SEC * 1000.0;
  printf("with_check:     %8.2f ms  sum=%d\n", ms1, (int)s1);
  printf("without_check:  %8.2f ms  sum=%d\n", ms2, (int)s2);
  return 0;
}

File: bench/s01_bounds_check.zig
// s01_bounds_check.zig — bounds check on/off overhead
// Compares "with bounds check" vs "without bounds check" array access.
// Matches bench/s01_bounds_check.c / .x algorithm.
// N=10000000 elements, R=10 rounds.
const std = @import("std");

const N: usize = 10_000_000;
const R: i32 = 10;
const N_I32: i32 = 10_000_000;

var arr: [N]i32 = undefined;
var idx: [N]i32 = undefined;

fn bench_with_check() i32 {
    var sum: i32 = 0;
    var r: i32 = 0;
    while (r < R) : (r += 1) {
        var i: usize = 0;
        while (i < N) : (i += 1) {
            const ii = idx[i];
            if (ii >= 0 and ii < N_I32) {
                sum +%= arr[@intCast(ii)];
            }
        }
    }
    return sum;
}

fn bench_without_check() i32 {
    var sum: i32 = 0;
    var r: i32 = 0;
    while (r < R) : (r += 1) {
        var i: usize = 0;
        while (i < N) : (i += 1) {
            sum +%= arr[@intCast(idx[i])];
        }
    }
    return sum;
}

pub fn main(init: std.process.Init) !void {
    var i: usize = 0;
    while (i < N) : (i += 1) {
        arr[i] = @intCast(i);
        const idx_val = @as(i32, @intCast(i)) *% 1103515245 +% 12345;
        idx[i] = @mod(idx_val, N_I32);
    }

    const ts0 = std.Io.Timestamp.now(init.io, .awake);
    const s1 = bench_with_check();
    const ts1 = std.Io.Timestamp.now(init.io, .awake);
    const s2 = bench_without_check();
    const ts2 = std.Io.Timestamp.now(init.io, .awake);

    const ns1: i64 = @intCast(ts0.durationTo(ts1).nanoseconds);
    const ns2: i64 = @intCast(ts1.durationTo(ts2).nanoseconds);

    const ms1 = @as(f64, @floatFromInt(ns1)) / 1_000_000.0;
    const ms2 = @as(f64, @floatFromInt(ns2)) / 1_000_000.0;

    std.debug.print("with_check:     {d:.2} ms  sum={d}\n", .{ ms1, s1 });
    std.debug.print("without_check:  {d:.2} ms  sum={d}\n", .{ ms2, s2 });
}
